### ISSUE TITLE
Define which docker image to use in build

### ISSUE THREAD

**Issue by ieguinoa** [databio/refgenie#175]
Hi,

I was wondering if it would make sense to have an option to have specify the container to use while calling build, this would allow to use different images for different build recipes and even for the same recipe at different times, which is a must if we want to create indexes for different versions of a tool.

I can help with coding the feature, just was not sure if it's in the plans.

cheers,
Ignacio

**Comment by stolarczyk** [comment]
hi @ieguinoa,

we're in the process of switching from monolithic containers approach to modular, multi-container environments. Check out [`bulker`](http://bulker.io/) documentation, which describes the concept in detail if you're interested :) That said, I'm not sure if it makes sense to invest in development of such a feature.

**Comment by nsheff** [comment]
Well, actually, isn't it the case that you can already specify the docker container for the recipe? If not I think this will definitely have to be a part of the flexible recipe format.

**Comment by ieguinoa** [comment]
Indeed, I see now that the container can be set in the recipe new recipes. It's defined as CONT: "databio/refgenie" for most of them, and from the documentation I understood that -d option always ended up running things inside databio/refgenie container, but it can be set for other recipes. 
Something related with this, since the version of the tool is fixed to the container and therefor to the recipe. Should't the asset name contain the version as part of the id? or is it possible to include a default (fixed) tag attached to the container being used that would indicate the version of the tool.

**Comment by nsheff** [comment]
> but it can be set for other recipes.

Indeed. Since we controlled all the recipes if they are built-in, we just made sure the `databio/refgenie` container had access to everything that was required to build any recipe, and then just used that for all recipes. But yes, it is more flexible really,  and the long-term goal was that the container would become more useful with flexible recipes.

> Should't the asset name contain the version as part of the id? or is it possible to include a default (fixed) tag attached to the container being used that would indicate the version of the tool.

it should rather be a tag, as you suggest. Right now the tag is set by the user, though, not by the recipe. The challenge we've had is how to automate this. you may want other things in the tag, like the particular flavor of the index for example. Our current thought is to include in the recipe some way to specify particular metadata about the run, which would be executed upon building. Some of this could be made available for a default tag.

We could call these variables 'metavars'. For example, if the asset is a salmon index, you'd add to the recipe.
```
metavars:
  version: 'salmon --version'
  type: "partial_sa_index"
```
Then you could use these in a default tag field:
```
default_tag: "{version}-{type}"
```
I guess it would be simplified if you just said `version: 1.3`, but by grabbing it as output from a command, which would run in the same environment as the recipe, allows us to:
1. log this information into the log file of the asset build
2. guarantee the version matches the tag, since it's generated by the software
3. use the same recipe without having to change anything other than the container, but getting different tags, reducing maintenance effort for the recipe

the disadvantage being that you have to have tools that can spit out their versions or whatnot. But I suppose if that's not the case you can always fall back to just encoding it like I did for 'type'. above


Related to #186

**Comment by ieguinoa** [comment]
I agree with the use of tags and I think the default_tag field is perfect.
But this would require the existence of a parameter to indicate which container to use, and I'm talking about a runtime parameter, in order to call the build command for the same asset but with different containers. This is the only way to allow multiple versions of an index to be run. 
Following this, to build different versions of a same 
```
refgenie build biocontainers/bowtie2_v1.8 hg38/bowtie2_index --assets fasta=hg38_cdna/fasta
refgenie build -d biocontainers/bowtie2_v2.3.5.1 hg38/bowtie2_index --assets fasta=hg38_cdna/fasta
```

The way it works now,  -d is a flag that always runs the recipe commands in the container defined in the recipe. 
It's something to discuss further though, I'm not sure how frequently one would like to create a new index for a new version. In the case of salmon for example I know they tend to change the versions quite frequently and the indexes are not backwards compatible. 

In any case, if the tool changes the interface to create the index then a new recipe needs to be added, but that shouldn't be the common case.

**Comment by illusional** [comment]
I'm interested in something like this too and happy to help find and annotate some containers for common index types. Is there a PR for following `bulker` support, keen to use this on HPCs, but am not allowed to run Docker. If there are small issues I'm interested in approaching some of them.

Also, I'm getting issues building indexes on macOS using the monolithic refgenie container, because of mounts denied (`/etc/sudoers.d`). It does feel like it's requesting a lot of mounts for building a `bwa_index` though.

**Comment by stolarczyk** [comment]
> I'm interested in something like this too and happy to help find and annotate some containers for common index types. Is there a PR for following bulker support, keen to use this on HPCs, but am not allowed to run Docker. If there are small issues I'm interested in approaching some of them.

 - [refgenie bulker manifest](https://github.com/databio/hub.bulker.io/blob/master/databio/refgenie_0.7.6.yaml) -- a set of containers that we use to build the assets
-  [tutorial](https://bulker.databio.org/en/latest/refgenie_tutorial/) with usage instructions

> Also, I'm getting issues building indexes on macOS using the monolithic refgenie container, because of mounts denied (/etc/sudoers.d). It does feel like it's requesting a lot of mounts for building a bwa_index though.

Did you see this issue? https://github.com/databio/bulker/issues/56#issuecomment-677708788

**Comment by nsheff** [comment]
Hey @illusional and everyone,

The idea of tagging images based on build came up again in #195. I'm trying to merge these ideas into a proposal for a new, more flexible recipe format, which is presented in issue #198. I'd love to get your feedback there; that's a bigger-picture issue that includes and transcends this issue, but would solve this as well. 

> Also, I'm getting issues building indexes on macOS using the monolithic refgenie container,

I bet it's due to this: https://github.com/databio/pypiper/blob/67908f2ee5f51fa5fdddb67eb6d7891aefeeda6a/pypiper/manager.py#L1904-L1908

It's doing that so that it can use the current host user, instead of the root container user, so that the build indexes are owned by the user instead of root -- but as I recall docker on MacOS doesn't behave the same way and these are not necessary there. So, I think this is a separate issue, which is that building @stolarczyk posted a workaround, but rather than this, we should probably just allow the pypiper docker feature to work on MacOS without a workaround. I've created an issue in pypiper to track this: https://github.com/databio/pypiper/issues/172

**Comment by nsheff** [comment]
Referring to back to an earlier comment by @ieguinoa in this thread: 

> I'm talking about a runtime parameter, in order to call the build command for the same asset but with different containers. 

I see what you mean. This is an interesting idea. Once we have #195, do you think this could be accomplished with writing a new recipe that points to that image? Or is that overkill? Is there anything else you'd want to adjust on-the-fly from the CLI?

I think I like the idea of making the image a CLI-adjustable parameter, but which defaults to the one in the recipe.

**Comment by illusional** [comment]
Thanks for the reply @stolarczyk and @nsheff, can confirm that exporting `/etc/sudoers.d` in my docker config allowed it to proceed. I also better understand how `bulker` is supposed to work now. Basically, it creates an alias for specific commands, like it (effectively) aliases `bwa` with `docker run <bwa container>`. So you load the refgenie bulker crate (for example):

```
pip install bulker
bulker load databio/refgenie:0.7.0
bulker activate databio/refgenie:0.7.0
refgenie build hg38/bowtie2_index
```

And then when refgenie calls `bwa index` or `bowtie2-build`, bulker has aliased this command (for docker / singularity). This means you DON'T provide the `-d` option and you don't end up with the monolithic build container.
